src/wp-includes/class-wp-autoload.php

```diff
@@ -0,0 +1,516 @@
+		'wp_privacy_requests_table'                   => 'wp-admin/includes/class-wp-privacy-requests-table.php',
+		'wp_screen'                                   => 'wp-admin/includes/class-wp-screen.php',
+		'wp_site_health_auto_updates'                 => 'wp-admin/includes/class-wp-site-health-auto-updates.php',
+		'wp_site_health'                              => 'wp-admin/includes/class-wp-site-health.php',
+		'wp_site_icon'                                => 'wp-admin/includes/class-wp-site-icon.php',
+		'wp_terms_list_table'                         => 'wp-admin/includes/class-wp-terms-list-table.php',
+		'wp_theme_install_list_table'                 => 'wp-admin/includes/class-wp-theme-install-list-table.php',
+		'wp_themes_list_table'                        => 'wp-admin/includes/class-wp-themes-list-table.php',
+		'wp_upgrader_skin'                            => 'wp-admin/includes/class-wp-upgrader-skin.php',
+		'wp_upgrader'                                 => 'wp-admin/includes/class-wp-upgrader.php',
+		'wp_users_list_table'                         => 'wp-admin/includes/class-wp-users-list-table.php',
+
+		/* Classes in wp-admin/includes/deprecated.php. */
+		'wp_user_search'                              => 'wp-admin/includes/deprecated.php',
+		'wp_privacy_data_export_requests_table'       => 'wp-admin/includes/deprecated.php',
+		'wp_privacy_data_removal_requests_table'      => 'wp-admin/includes/deprecated.php',
+	);
+
+	/**
+	 * Whether the autoloader has already been registered or not.
+	 *
+	 * Avoid registering the autoloader multiple times.
+	 *
+	 * @static
+	 * @access private
+	 *
+	 * @var bool
+	 */
+	private static $registered = false;
+
+	/**
+	 * Register the autoloader.
+	 *
+	 * @return void
+	 */
+	public static function register() {
+		// Bail early if already registered.
+		if ( static::$registered ) {
+			return;
+		}
+
+		static::register_external_bundled();
+		static::register_core();
+
+		static::$registered = true;
+	}
+
+	/**
+	 * Register the autoloader for external, bundled libraries.
+	 *
+	 * @return void
+	 */
+	public static function register_external_bundled() {
+		require_once ABSPATH . 'wp-includes/Requests/src/Autoload.php';
+		require_once ABSPATH . 'wp-includes/sodium_compat/autoload.php';
+
+		spl_autoload_register( array( '\WpOrg\Requests\Autoload', 'load' ) );
+	}
+
+	/**
+	 * Register the autoloader for WordPress Core classes.
+	 *
+	 * @return void
+	 */
+	public static function register_core() {
+		spl_autoload_register( array( __CLASS__, 'autoload_core' ), true, true );
+	}
+
+	/**
+	 * Autoload a WordPress class.
+	 *
+	 * @param string $class Class name.
+	 * @return bool True if the class was loaded, false otherwise.
+	 */
+	public static function autoload_core( $class_name ) {
+		// Lowercase the classname to accomodate for WP classes written with wrong cases.
+		$class_name = strtolower( $class_name );
```

Comment:
Are WP classes written with wrong cases currently? Shouldn't this be fixed wherever it is done, while this is now hiding this issue -- and adds another function call?

Reply:
You'd be shocked the number of times I see `WP_QUERY` or `wp_Query` in support questions, not including this will probably break a lot of code/sites, not that they should ever been able to work to begin with having misspelt classes

Reply:
>  Shouldn't this be fixed wherever it is done, while this is now hiding this issue -- and adds another function call?

I agree with @TobiasBg that these should be fixed. I had a similar issue with a different project, autoloading classes with the wrong casing. The solution I used was to register two autoload callbacks: the first one was with the correct casing, and the second one was the case-insensitive one. The second one does `trigger_error()` each time a match is found with the wrong casing. With this approach, we can continue working with the incorrectly cased classes, and at the same time warn developers about the incorrect casing they have used for the classes.

Reply:
With the size of the WP ecosystem and the number of plugins/themes etc out there, this would probably trigger lots of errors in many sites... 
Using the wrong case when invoking classes - though ill-advised - is technically not an error. If at some point PHP itself changes this, then we could change the behaviour of the autoloader, but at this point in time I don't see how this would benefit the ecosystem 🤔

Reply:
> Using the wrong case when invoking classes - though ill-advised - is technically not an error.

@aristath That's right, it is not an error. That's why my suggestion is to use `trigger_error()` as this can be reported as E_USER_WARNING or E_USER_NOTICE, or even E_USER_DEPRECATED.

Reply:
What I meant is, I don't know if we should.
Since this is not _wrong_, it's a preference. I don't like it when I see devs using what I consider "wrong" case, but they have every right to do so. It's a personal preference and not something that is universally considered "wrong". It is not our place to try and enforce our opinion to others on how they should write code - no matter how justified we feel our opinion might be. It's not an error, PHP allows it (so far), therefore we should just accommodate it.

Since it's not actually wrong, it's not a matter of code. Instead, it turns into a matter of politics and ethics.... And I just don't want to get involved in discussions like that. At least not in this PR... I'm trying to keep this PR as simple, non-invasive and politics-agnostic as possible. Ideally we can get this PR merged in Core sometime in the next 1-2 major releases, and then we can improve and/or tweak it based on community feedback 😅

Reply:
Note that enforcing correct capitalisation would be a coding standard added on top of PHP, therefore it would be a better fit for WPCS if we wanted to go down that route. The most I'd ever want to see here related to this is a warning, never an error. `trigger_error` would have major consequences for already working code

Reply:
@aristath @tomjn thanks for exploring this topic with me. Good luck with this patch